src/bin/pg_autoctl/supervisor.c

```diff
@@ -56,22 +72,277 @@ static bool supervisor_restart_service(Supervisor *supervisor,
 									   Service *service,
 									   int status);
 
+static void supervisor_handle_failed_restart(Supervisor *supervisor,
+											 Service *service,
+											 int status);
+
 static bool supervisor_may_restart(Service *service);
 
 static bool supervisor_update_pidfile(Supervisor *supervisor);
 
+static int supervisor_handle_dynamic(Supervisor *supervisor);
+
+
+/*
+ * supervisor_it_* implement a helper to loop through the active services in a
+ * supervisor struct. They can be used either on their own, or via the helper
+ * macro foreach_supervised_service().
+ * Consult the macro for an example of use.
+ */
+static inline Service *
+supervisor_it_get(Supervisor_it *sit)
+{
+	if (sit->index < sit->count)
+	{
+		return &sit->services[sit->index];
+	}
+
+	return NULL;
+}
+
+
+static inline Service *
+supervisor_it_first(Supervisor_it *sit, Supervisor *supervisor)
+{
+	*sit = (Supervisor_it) {
+		.supervisor = supervisor,
+		.services = supervisor->services,
+		.count = supervisor->serviceCount,
+		.index = 0,
+	};
+
+	return supervisor_it_get(sit);
+}
+
+
+static inline Service *
+supervisor_it_next(Supervisor_it *sit)
+{
+	sit->index++;
+
+	/* if we are done with the static services, move to the dynamic */
+	if (sit->index >= sit->count &&
+		sit->services == sit->supervisor->services)
+	{
+		sit->services = sit->supervisor->dynamicServices.services;
+		sit->count = sit->supervisor->dynamicServices.serviceCount;
+		sit->index = 0;
+	}
+
+	return supervisor_it_get(sit);
+}
+
+
+#define foreach_supervised_service(it, supervisor, service) \
+	for ((service) = supervisor_it_first((it), (supervisor)); \
+		 (service) != NULL; \
+		 (service) = supervisor_it_next(it))
+
+
+/*
+ * supervisor_add_dynamic starts the service and if successfull, adds it to the
+ * dynamicService array.
+ *
+ * Since the services are largely discoverable by name, the name has to be
+ * unique. If it is not, then the service is not started and the function
+ * returns false.
+ * Also since the service runs under supervision, there must be space for it to
+ * be added to the supervisor's dynamic services array. If there is not, then
+ * the service is not started and the function returns false.
+ * The service will not be added if a signal other than "dynamic service" has
+ * been received.
+ *
+ * Returns true if successfully starts the service.
+ */
+bool
+supervisor_add_dynamic(Supervisor *supervisor, Service *service)
+{
+	Service *dynamicService;
+	sigset_t sig_mask;
+	sigset_t sig_mask_orig;
+	int serviceIndex = supervisor->dynamicServices.serviceCount;
+	bool started;
+
+	if (supervisor_service_exists(supervisor, service->name))
+	{
+		log_error("Service %s already exists under supervision",
+				  service->name);
+		return false;
+	}
+
+	/* 0-based */
+	if (serviceIndex >= (MAXDYNSERVICES - 1))
+	{
+		log_error("Reached maximum permitted number of dynamic services, "
+				  "service %s is not added",
+				  service->name);
```

Comment:
Our error messages begin with "Failed to" which often makes it easier for the user to grasp the context of the error and its impact on the operations. Something like "Failed to start dynamic service \"%s\": pg_autoctl supports up to %d dynamic services and all the slots are used already".

Reply:
Great! Will amend!